File: sort/quick/application/topK.go
/*
 * @Author: yangxcc
 * @version: 1.0
 * @Date: 2022-12-21 14:26:08
 * @LastEditTime: 2022-12-26 16:34:22
 */
package application

// TopK问题，在升序排列的数组中，第k大元素的位置在len(nums)-k
func findKthLargest(nums []int, k int) int {
	left, right := 0, len(nums)-1

	for left <= right {
		// 在[left, right]范围内找pivot
		pivot := partition(nums, left, right)
		if pivot == len(nums)-k {
			return nums[pivot]
		} else if pivot < len(nums)-k {
			// 区间调整为[pivot+1, right]
			left = pivot + 1
		} else {
			// 区间调整为[left, pivot-1]
			right = pivot - 1
		}
	}

	return -1
}

func partition(nums []int, left, right int) int {
	pivot := nums[left]

	leftIdx, rightIdx := left+1, right
	for leftIdx <= rightIdx {
		for leftIdx < right && nums[leftIdx] <= pivot {
			leftIdx++
		}

		for rightIdx > left && nums[rightIdx] > pivot {
			rightIdx--
		}

		if leftIdx >= rightIdx {
			break
		}

		nums[leftIdx], nums[rightIdx] = nums[rightIdx], nums[leftIdx]
	}

	nums[left], nums[rightIdx] = nums[rightIdx], nums[left]

	return rightIdx
}

File: sort/quick/application/topK.java
/*
 * @Author: yangxcc
 * @version: 1.0
 * @Date: 2022-12-21 14:26:14
 * @LastEditTime: 2023-02-01 17:13:20
 */
package sort.quick.application;

/**
 * leetcode 215 middle 数组中的第k个最大元素
 * 
 * 给定整数数组 nums 和整数 k，请返回数组中第 k 个最大的元素。
 * 请注意，你需要找的是数组排序后的第 k 个最大的元素，而不是第 k 个不同的元素。
 * 你必须设计并实现时间复杂度为 O(n) 的算法解决此问题。
 */
public class topK {
    public int findKthLargest(int[] nums, int k) {
        if (k > nums.length) {
            return -10001;
        }

        int left = 0, right = nums.length - 1;
        while (left <= right) {
            int position = partition(nums, left, right);
            if (position == nums.length - k) {
                return nums[position];
            } else if (position < nums.length - k) {
                left = position + 1;
            } else {
                right = position - 1;
            }
        }
        return -10001;
    }

    public int partition(int[] nums, int left, int right) {
        int pivot = nums[left];
        // 如下这么定义leftIdx和rightIdx，相当于是一个开区间，即
        // [left, leftIdx) <= povot, (rightIdx, right] > pivot
        int leftIdx = left + 1, rightIdx = right;
        // 因此这里是 <= ，这里吃过亏
        while (leftIdx <= rightIdx) {
            while (leftIdx < right && nums[leftIdx] <= pivot) {
                leftIdx++;
            }
            while (rightIdx > left && nums[rightIdx] > pivot) {
                rightIdx--;
            }

            if (leftIdx >= rightIdx) {
                break;
            }

            swap(nums, leftIdx, rightIdx);
        }

        swap(nums, left, rightIdx);
        return rightIdx;
    }

    public void swap(int[] nums, int L, int R) {
        int tmp = nums[L];
        nums[L] = nums[R];
        nums[R] = tmp;
    }
}
